### PR TITLE
adds docs about control flow primitives

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/3196.

### CODE REVIEW

File: docs/control_flow_primitives.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+stop processing here it is good enough to simply `return` within the code. The
+resources being configured after the canceled resource are then executed and the
+canceled resource is executed again on the next reconciliation loop once the
+resync period kicks in again.
+
+#### CRUD Resources
+
+CRUD resources implement a couple of steps to help with structuring more complex
+use cases. These steps act like a framework and help navigating through the
+stormy waters of reconciliation and its implications. In order to cancel
```
Comment (kopiczko) [inline docs/control_flow_primitives.md]: LoL creativity :D

File: docs/control_flow_primitives.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+## Cancel Reconciliation
+
+In order to cancel the whole reconciliation you can simply call
+[`reconciliationcanceledcontext.SetCanceled(ctx)`](https://github.com/giantswarm/kvm-operator/blob/de7e109f4a652b785bbcf4214a1c8e028bf0eed4/service/controller/v12/resource/namespace/current.go#L45)
+which will then stop executing all configured resources within the current
+reconciliation loop. On the next reconciliation loop all resources are executed
+again based on how they got configured.
+
+## Repeat Delete Events
```
Comment (kopiczko) [inline docs/control_flow_primitives.md]: `Repeat Cancelled Delete Events`?

File: docs/control_flow_primitives.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+A convention for CRUD resource cancelation is to do it within `GetCurrentState`
+when possible.
+
+## Cancel Reconciliation
+
+In order to cancel the whole reconciliation you can simply call
+[`reconciliationcanceledcontext.SetCanceled(ctx)`](https://github.com/giantswarm/kvm-operator/blob/de7e109f4a652b785bbcf4214a1c8e028bf0eed4/service/controller/v12/resource/namespace/current.go#L45)
+which will then stop executing all configured resources within the current
+reconciliation loop. On the next reconciliation loop all resources are executed
+again based on how they got configured.
```
Comment (kopiczko) [inline docs/control_flow_primitives.md]: I would add


```Cancelling deletion event reconciliation will cause the finalizer deletion. This means the deletion event won't be replayed. When this behavior is not desired check [Repeat Delete Events](#repeat-delete-events).```

Comment (rossf7) [inline docs/control_flow_primitives.md]: +1 Also 🇬🇧 `were configured`

File: docs/control_flow_primitives.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+## Repeat Delete Events
+
+There are separate docs about [using finalizers](using_finalizers.md) which
+describe a lot of background. Thus we just touch the control flow aspects of
```
Comment (rossf7) [inline docs/control_flow_primitives.md]: of the background

File: docs/control_flow_primitives.md (hunk 1 above)
Comment (xh3b4sd) [reply docs/control_flow_primitives.md]: For real man. <3

File: docs/control_flow_primitives.md (hunk 2 above)
Comment (xh3b4sd) [reply docs/control_flow_primitives.md]: They do not get canceled necessarily.

File: docs/control_flow_primitives.md (hunk 3 above)
Comment (xh3b4sd) [reply docs/control_flow_primitives.md]: I added the information in a reworded version.

Comment (xh3b4sd) [reply docs/control_flow_primitives.md]: Done.

File: docs/control_flow_primitives.md (hunk 4 above)
Comment (xh3b4sd) [reply docs/control_flow_primitives.md]: Done.

**Pull request merged**